mypy/checker.py

```diff
@@ -1904,16 +1904,22 @@ def check_assignment(self, lvalue: Lvalue, rvalue: Expression, infer_lvalue_type
                 self.check_forbidden_inference_types(lvalue, rvalue, rvalue_type)
                 self.infer_variable_type(inferred, lvalue, rvalue_type, rvalue)
 
-    def check_forbidden_inference_types(self, lvalue: Lvalue, rvalue: Context, rvalue_type: Type) -> None:
-        if (isinstance(rvalue_type, Instance) and isinstance(lvalue, RefExpr)):
+    def check_forbidden_inference_types(self, lvalue: Lvalue, rvalue: Context,
+                                        rvalue_type: Type) -> None:
+        """Check for any list or dict literals that are inferred ambiguously and require annotation.
+        """
+        if not self.scope.top_function() and self.options.allow_untyped_globals:
+            return
+
+        if isinstance(rvalue_type, Instance) and isinstance(lvalue, RefExpr):
             if isinstance(rvalue, ListExpr):
                 list_arg_type = rvalue_type.args[0]
                 if isinstance(list_arg_type, Instance):
                     if list_arg_type.type.fullname() != 'builtins.object':
                         # This list has a specific type, no further checks needed.
                         return
 
-                    if self.iterable_has_object_type(iter(rvalue.items)):
+                    if self.iterable_has_object_type(rvalue.items):
                         return
 
                     self.msg.need_annotation_for_var(lvalue.node, rvalue)
```

Comment:
This currently throws a mypy error:
```
mypy/checker.py:1919: error: Argument 1 to "need_annotation_for_var" of "MessageBuilder" has incompatible type "Optional[SymbolNode]"; expected "SymbolNode"
```

I guess I need to check if lvalue.node is none, but what do I do if it is?

Reply:
I think the best way to solve this, is to use `inferred` as the first argument of this method, which is a non-optional `Var` at the place where the method is called  (instead of `lvalue`). You only use `lvalue` once in an `if` at the top, instead you can move this `if` to the call site (this will also save one indentation level here).

Reply:
According to my read, `inferred` is set by `check_lvalue` which returns a `Tuple[Optional[Type], Optional[IndexExpr], Optional[Var]]`, so we still have to check if the `Var` is there. Perhaps I am missing something?

For moving the `if`, you're talking about `if isinstance(rvalue_type, Instance) and isinstance(lvalue, RefExpr):`, correct?

Reply:
> According to my read, inferred is set by `check_lvalue` which returns a `Tuple[Optional[Type], Optional[IndexExpr], Optional[Var]]`

Well your reading is wrong. Please read more carefully, there is an unpacking.
> For moving the `if`, you're talking about `if isinstance(rvalue_type, Instance) and isinstance(lvalue, RefExpr):`, correct?

Yes.